lib/core/src/Cardano/Wallet/DaedalusIPC.hs

```diff
@@ -0,0 +1,231 @@
+        sayErr $ "[ERROR] Starting Daedalus IPC: " <> err
+  where
+    -- How to respond to an incoming message, or when there is an incoming
+    -- message that couldn't be parsed.
+    msg (Right QueryPort) = Just (ReplyPort port)
+    msg (Left e) = Just (ParseError e)
+
+    -- What to do in context of withNodeChannel
+    action :: (MsgOut -> IO ()) -> IO ()
+    action send = send Started >> sleep
+
+    sleep = threadDelay maxBound
+
+----------------------------------------------------------------------------
+-- NodeJS child_process IPC protocol
+-- https://nodejs.org/api/child_process.html#child_process_child_process_spawn_command_args_options
+
+data NodeChannelError
+    = NodeChannelDisabled
+      -- ^ This process has not been started as a nodejs @'ipc'@ child_process.
+    | NodeChannelBadFD Text
+      -- ^ The @NODE_CHANNEL_FD@ environment variable has an incorrect value.
+    deriving (Show, Eq)
+
+withNodeChannel
+    :: (FromJSON msgin, ToJSON msgout)
+    => (Either Text msgin -> IO (Maybe msgout))
+       -- ^ Incoming message handler
+    -> ((msgout -> IO ()) -> IO a)
+       -- ^ Action to run
+    -> IO (Either NodeChannelError (IO ()))
+withNodeChannel onMsg handleMsg = fmap setup <$> lookupNodeChannel
+  where
+    setup handle = do
+        chan <- newEmptyMVar
+        let ipc = ipcListener handle onMsg chan
+            action' = handleMsg (putMVar chan)
+        race_ action' ipc
+
+-- | Parse the NODE_CHANNEL_FD variable, if it's set, and convert to a
+-- 'System.IO.Handle'.
+lookupNodeChannel :: IO (Either NodeChannelError Handle)
+lookupNodeChannel = (fromMaybe "" <$> lookupEnv "NODE_CHANNEL_FD") >>= \case
+    "" -> pure (Left NodeChannelDisabled)
+    var -> case readEither var of
+        Left err -> pure . Left . NodeChannelBadFD $
+           "unable to parse NODE_CHANNEL_FD: " <> T.pack err
+        Right fd -> tryJust handleBadFd (fdToHandle fd)
+  where
+    handleBadFd :: IOException -> Maybe NodeChannelError
+    handleBadFd = Just . NodeChannelBadFD . T.pack . show
+
+ipcListener
+    :: forall msgin msgout. (FromJSON msgin, ToJSON msgout)
+    => Handle
+    -> (Either Text msgin -> IO (Maybe msgout))
+    -> MVar msgout
+    -> IO ()
+ipcListener handle onMsg chan = do
+    hSetNewlineMode handle noNewlineTranslation
+    catch (race_ replyLoop sendLoop) onIOError
+  where
+    sendLoop, replyLoop :: IO ()
+    replyLoop = forever (recvMsg >>= onMsg >>= maybeSend)
+    sendLoop = forever (takeMVar chan >>= sendMsg)
+
+    recvMsg :: IO (Either Text msgin)
+    recvMsg = first T.pack . eitherDecode <$> readMessage handle
+
+    sendMsg :: msgout -> IO ()
+    sendMsg = sendMessage handle . encode
+
+    maybeSend :: Maybe msgout -> IO ()
+    maybeSend = maybe (pure ()) (putMVar chan)
+
+    onIOError :: IOError -> IO ()
+    onIOError err = do
+      sayErrString $ "[ERROR] Exception caught in DaedalusIPC: " <> show err
+      when (isEOFError err) $ sayErr "[DEBUG] it's an eof"
```

Comment:
An `EOF` here would mean that Daedalus / the client has closed the connection on their hand, right ?

Reply:
Usually the exception is usually broken pipe. I think this was debug code that got left in cardano-sl from the DEVOPS-1234 investigation. I have removed the log anyway.